Wangscape/utils.h

```diff
@@ -0,0 +1,64 @@
+#pragma once
+#include <type_traits>
+#include <stdexcept>
+
+namespace
+{
+
+template<typename I>
+I _pow(I base, I exp)
+{
+    I result = 1;
+    for (int i = 0; i < std::numeric_limits<I>::digits; i++)
+    {
+        if (exp & 1)
+            result *= base;
+        else if (exp == 0)
+            break;
+        exp >>= 1;
```

Comment:
The main reason to use `/` and `%` here is compatibility with custom types that might not have `>>`. In that case, it would probably best to add a `bool UseBitshift=true` to the template arguments so that the special cases for `base == 2` and `base == -2` can be turned off.